index.d.ts

```diff
@@ -16,6 +20,46 @@ export interface Options {
 	readonly stopOnError?: boolean;
 }
 
+export interface OngoingMappingsStopOptions {
+	/**
+	Whether or not to remove all holes from the result array (caused by pending mappings).
+
+	@default false
+	*/
+	readonly collapse?: boolean;
+
+	/**
+	Value to use as immediate result for pending mappings, replacing holes from the result array.
+
+	This option is ignored if `collapse` is set to `true`.
+
+	@default undefined
+	*/
+	readonly fillWith?: unknown;
+}
+
+export interface StopOptions<NewElement> {
+	/**
+	Value to provide as result for this iteration.
+
+	@default undefined
+	*/
+	readonly value?: NewElement;
+
+	/**
+	Options to configure what `pMap` must do with any concurrent ongoing mappings at the moment `stop` is called.
+	*/
+	readonly ongoingMappings?: OngoingMappingsStopOptions;
+}
+
+type BaseStopValueWrapper<NewElement> = {
+	[stop]: Required<StopOptions<NewElement>>;
+};
+
+export type StopValueWrapper<NewElement> = NewElement extends any ? BaseStopValueWrapper<NewElement> : never;
+
+type MaybeWrappedInStop<NewElement> = NewElement | StopValueWrapper<NewElement>;
```

Comment:
Why these 3 types? Not immediately clear why this complication is needed. We don't need to expose anything about `stop` to the user in the return value. For the user, it can just be an opaque object.

Reply:
Do you have any favorite way of declaring an opaque object in TypeScript? Recenly I've seen `Record<never, never>`, does this sound good?

Reply:
> We don't need to expose anything about stop to the user in the return value. For the user, it can just be an opaque object.

Actually, on second thought, I think we have to, because as soon as the user uses `return pMap.stop()` inside `mapper`, TypeScript will infer the return type of `mapper` differently, and we need TS to infer it to something that we can unwrap later in order to provide a correct return type for the `pMap` call itself. This was quite complicated to achieve to be honest. I don't see how it would be possible if we simplify the result of `.stop` to a constant type such as `Record<never, never>`.

Let's merge it this way, currently, and you can open an issue to improve typings? Then if someone finds a way to do it without breaking the current TS tests I added, it can be done later. What do you think?

Reply:
Can't you just do what was done here: https://github.com/sindresorhus/p-map/commit/c9c088283e07280fb2c49ed4f83fd20cc0ad8b46# (Exclude the symbol from the return type)

// @Richienb

Reply:
When a stop value is recieved, it could set a variable which causes future resolutions of other items